### ISSUE TITLE
关于输出结果中predict_score和gt_score的疑问

### ISSUE THREAD

**Issue by andyjiang1116** [zzc-1998/Q-Eval#5]
您好，我有两个疑问：
第一：`infer.py`中输出结果打印的`predict_score`的数值是0到1的值，怎么对应到最终的评分结果呢，评分结果按理说应该是在1到5之间才能和最后的输出对应吧？
第二：`predict_score`如果是0到1的值，怎么和标注中的`gt_score`做loss计算呢，因为`gt_score`在标注文件里是1到5之间的值。

**Comment by zzc-1998** [comment]
你好，

第一个可以参考这个函数：

def wa5(logits):
    import numpy as np
    logprobs = np.array([logits["Excellent"], logits["Good"], logits["Fair"], logits["Poor"], logits["Bad"]])
    probs = np.exp(logprobs) / np.sum(np.exp(logprobs))
    return np.inner(probs, np.array([1,0.75,0.5,0.25,0.]))

infer输出分数的具体做法是根据五个形容词的概率，最后通过加权平均输出分数的，所以具体的分数范围只和这个函数有关，你只需要调整权重参数就可以调整输出范围。同时这个函数是线性，不会影响输出分数的排序关系。

第二个问题：

q-eval采用的交叉熵损失函数，1-5的gt-score会被转化为形容词并且组成sft的问答对来监督大模型。所以gt_score不会直接参与loss的计算。

**Comment by andyjiang1116** [comment]
感谢您的解答，但是这边还是不太理解
第一：模型输出的logits加权后求平均，这个怎么映射对应到最终的结果呢？比如结果是Good还是Bad？
第二：gt_score转化为形容词这个我明白，但是论文中提到还有个MSELoss计算gt_score和predict_score，这部分训练的时候是怎么实现的呢？

**Comment by zzc-1998** [comment]
第一：如果你是想要一个最终的质量分类结果，那么并不需要做这个加权平均，直接取概率最大的形容词最为最终结果即可。这个加权平均是为了适应质量评价分数量化的需求。

第二：MSELoss的计算只需要要在训练中把分数统一到同一个尺度就可以，你可以选择把predict_score映射到1-5，也可以选择把gt_score映射到0-1。

**Comment by andyjiang1116** [comment]
好的，感谢解答，也就是说第一个问题中，最终模型输出的`logits`的最后一个字符（对应代码中`["logits"][:,-1]`）是从5类中选取的最大概率作为最终结果是吗？

**Comment by zzc-1998** [comment]
["logits"][:,-1] 是代表了模型输出的最后一个单词的字典形式的词概率向量，里面包含了所有词典里单词的概率。我们从中索引了五个形容词的概率来参与质量分数的计算。如果你只是想要分类，应该是选择索引五个形容词之后最大的概率对应的那个形容词，也就是max（[logits["Excellent"], logits["Good"], logits["Fair"], logits["Poor"], logits["Bad"]]）

**Comment by andyjiang1116** [comment]
好的，谢谢。那第二个问题中训练阶段计算MSELoss的时候，gt_score可以映射到0到1之间，那么对应的predict_score是怎么获取的呢？是wa5函数的计算方式吗？还是说取这5个单词的softmax之后的最大概率？

**Comment by zzc-1998** [comment]
计算MSELoss的时候，预测分数由wa5函数获得。单词概率是参与的交叉熵损失函数。

**Comment by andyjiang1116** [comment]
那可以通过将gt_score映射到0到1之间和wa5函数计算后的模型预测分数来计算MSELoss是吗？

**Comment by zzc-1998** [comment]
可以的

**Comment by andyjiang1116** [comment]
感谢您的解答~
